libz-rs-sys/src/gz.rs

```diff
@@ -249,28 +288,166 @@ unsafe fn gzopen_help(path: *const c_char, fd: c_int, mode: *const c_char) -> gz
+        state.how = How::Look; // look for gzip header
+    } else {
+        state.reset = false; // no deflateReset pending
+    }
+    state.seek = false; // no seek request pending
+                        // Safety: It is valid to pass a null msg pointer to `gz_error`.
+    unsafe { gz_error(state, Z_OK, ptr::null_mut()) }; // clear error status
+    state.pos = 0; // no uncompressed data yet
+                   // FIXME add once the deflate state is implemented:
+                   // state->strm.avail_in = 0;       /* no input data yet */
+}
+
+// Set the error message for a gzip stream, and deallocate any
+// previously set error message.
+//
+// # Safety
+//
+// - `state` must be a properly constructed `GzState`, e.g. as produced by [`gzopen`]
+// - `msg` must be null or a valid C string. If `msg` is non-null, this function will
+//   make a copy if it, and the caller will retain ownership of the original.
+unsafe fn gz_error(state: &mut GzState, err: c_int, msg: *const c_char) {
+    if !state.msg.is_null() {
+        // NOTE: zlib-ng has a special case here: it skips the deallocation if
+        // state.err == Z_MEM_ERROR. However, we always set state.msg to null
+        // when state.err is set to Z_MEM_ERROR, so that case is unreachable
+        // here.
+        unsafe { deallocate_cstr(state.msg.cast_mut()) };
+        state.msg = ptr::null_mut();
+    }
+
+    // On error, set state.have to 0 so that the `gzgetc()` C macro fails
+    if err != Z_OK && err != Z_BUF_ERROR {
+        state.have = 0;
+    }
+
+    // Set the error code, and if no message, then done
+    state.err = err;
+    if msg.is_null() {
+        return;
+    }
+
+    // For an out of memory error, don't bother trying to allocate space for an error string.
+    // ([`gzerror`] will provide literal string as a special case for OOM errors.)
+    if err == Z_MEM_ERROR {
+        return;
+    }
+
+    // Format the error string to include the file path.
+    // Safety: `gzopen` and `gzdopen` ensure that `state.path` is a non-null C string,
+    //          the caller of this function is reponsible for ensuring that `msg` is a C string,
+    //          and we exit this function above if `msg` is null.
+    let sep = b": \0".as_ptr().cast::<c_char>();
+    let buf = &mut [0u8; 17];
+    state.msg = match state.source {
+        Source::Path(path) => unsafe { gz_strcat(&[path, sep, msg]) },
+        Source::Fd(fd) => unsafe { gz_strcat(&[fd_path(buf, fd).as_ptr(), sep, msg]) },
+    };
+
+    if state.msg.is_null() {
+        state.err = Z_MEM_ERROR;
+    }
+}
+
 // Deallocate a GzState structure and all heap-allocated fields inside it.
 //
 // # Safety
 //
 // - The `state` object and all heap-allocated fields within it must have been obtained
 //   using `ALLOCATOR`.
 // - The caller must not use the `state` after passing it to this function.
-unsafe fn free_state(state: &mut GzState) {
+unsafe fn free_state(state: *mut GzState) {
```

Comment:
this can't take a mutable reference (MIRI complains if it does. According to rust's model, you can't just destroy a mutable reference).